### PR TITLE
[clang] Fix CTAD not work for function-type and array-type arguments.

### LINKED ISSUE #51710: Class template argument deduction fails in case of function type argument
|  |  |
| --- | --- |
| Bugzilla Link | [52368](https://llvm.org/bz52368) |
| Version | trunk |
| OS | Windows NT |
| Reporter | LLVM Bugzilla Contributor |
| CC | @DougGregor,@zygoloid |

## Extended Description 
In the following program Clang is unable to perform class template argument deduction:
```
template<typename T>
struct A{
    A(T f()){ f(); }
};

int foo() { return 1; }

int main() {
    [[maybe_unused]] A y = foo; // Clang error
}
```
GCC accepts it successfully, demo: https://gcc.godbolt.org/z/PKe6T5aW5

Related discussion: https://stackoverflow.com/q/69778510/7325599

### PR DESCRIPTION
**Pull request by hokein**
Fixes https://github.com/llvm/llvm-project/issues/51710.

When transforming a constructor into a corresponding deduction guide, the decayed types (function/array type) were not handled properly which made clang fail to compile valid code. The patch teaches clang handle these decayed type in the transformation.

### CODE REVIEW

File: clang/test/SemaCXX/ctad-decay.cpp [hunk 1]
```diff
@@ -0,0 +1,26 @@
+// RUN: %clang_cc1 -fsyntax-only -verify -Wno-unused-value -std=c++20 %s
+// expected-no-diagnostics
```
Comment (cor3ntin) [inline clang/test/SemaCXX/ctad-decay.cpp:2]: can you move that test to clang/test/SemaCXX/cxx1z-class-template-argument-deduction.cpp ?

Comment (hokein) [reply clang/test/SemaCXX/ctad-decay.cpp]: Sure, done.

File: clang/test/SemaCXX/cxx1z-class-template-argument-deduction.cpp [hunk 2]
```diff
@@ -645,4 +645,27 @@ namespace undefined_warnings {
...
 }
+
+namespace GH51710 {
+template<typename T>
+struct A {
+  A(T f()) {}
+  A(int f(), T) {}
+
+  A(T array[10]) {}
+  A(int array[10], T) {}
```
Comment (cor3ntin) [inline clang/test/SemaCXX/cxx1z-class-template-argument-deduction.cpp:656]: Can you add tests for incomplete array (in another struct ie

```cpp
template<typename T>
struct B {
   B(T array[]) {}
   B(int array[], T) {}
};

File: clang/lib/Sema/SemaTemplate.cpp [hunk 3]
```diff
@@ -2587,15 +2587,17 @@ struct ConvertConstructorToDeductionGuideTransform {
...
-                                                NewDI->getType(),
-                                                NewDI,
-                                                OldParam->getStorageClass(),
-                                                NewDefArg.get());
+    // Handle arrays and functions decay.
+    auto NewType = NewDI->getType();
+    if (NewType->isArrayType())
+      NewType = SemaRef.Context.getArrayDecayedType(NewType);
+    else if (NewType->isFunctionType())
+      NewType = SemaRef.Context.getPointerType(NewType);
```
Comment (cor3ntin) [inline clang/lib/Sema/SemaTemplate.cpp:2595]: I think we can just use `SemaRef.getDecayedType(NewDI->getType())`, it covers both case

File: clang/test/SemaCXX/cxx1z-class-template-argument-deduction.cpp (hunk 2 above)
Comment (hokein) [reply clang/test/SemaCXX/cxx1z-class-template-argument-deduction.cpp:656]: Done.

File: clang/lib/Sema/SemaTemplate.cpp (hunk 3 above)
Comment (hokein) [reply clang/lib/Sema/SemaTemplate.cpp]: Good point, thanks. I was not aware of this API.

**Review by cor3ntin — APPROVED** [review]
LGTM, thanks

**Comment by shafik** [comment]
Can you please provide a more detailed summary, since these are usually what goes in the git log. It should describe the cause and the approach of the fix. It is also helpful for reviewers as well.

**Review by shafik — APPROVED** [review]
LGTM after more detailed summary provided.

**Comment by hokein** [comment]
> Can you please provide a more detailed summary, since these are usually what goes in the git log. It should describe the cause and the approach of the fix. It is also helpful for reviewers as well.

Done, added details in the description.

**Comment by shafik** [comment]
> > Can you please provide a more detailed summary, since these are usually what goes in the git log. It should describe the cause and the approach of the fix. It is also helpful for reviewers as well.
> 
> Done, added details in the description.

Thank you! Much appreciated :-)

**Pull request merged**